### ISSUE TITLE
Lots of requests to datapoints.json

### ISSUE THREAD

**Issue by adamwolf** [beeminder/BeeSwift#238]
Hi folks!

I was browsing some logs today and I noticed some access logs that looked a little funny. (I don't mean to imply this is causing any issues on the server.)

It looks like sometimes we'll get a user just slamming datapoints.json, like ~150 times in 8 seconds.

The requests all look like 

`GET /api/v1/users/USERNAME/goals/steps/datapoints.json?access_token=REDACTED&count=7&sort=daystamp HTTP/1.1"`

The beeminder API responds 304 to them, up until the last one.

It looks like it happens to different users, and for different goals for those users.

[question from krugerk about whether this was present in BeemiOS 5.7 or just appeared in 5.8]

**Comment by krugerk** [comment]
Thanks for opening the issue and letting the team know of an area for improvement.

The report thus far refers to the latest version, 5.8. Could we find the same issue in the previous version, 5.7? "BeeSwift/5.7" in the logs.

**Comment by krugerk** [comment]
There is only one place currently where the app [GETs datapoints.json](https://github.com/beeminder/BeeSwift/blob/e4700533c781e4e25352fbe53cb9064faa6d97ec/BeeSwift/JSONGoal.swift#L623) and that is in the healthkit integration part of JSONGoal.
That call does use [`.validate/()`](https://github.com/beeminder/BeeSwift/blob/e4700533c781e4e25352fbe53cb9064faa6d97ec/BeeSwift/RequestManager.swift#L17), which apparently expects the answer to be [within the 200s](https://github.com/Alamofire/Alamofire/blob/master/Documentation/AdvancedUsage.md#validation).


Maybe we need to do something app/client side to deal with the 304

**Comment by adamwolf** [comment]
Hi from 2022! This appears to still be an issue, but it's becoming more problematic.

I'm seeing this even with 200s.

```
200.107.24.229 - - [01/Oct/2022:14:14:58 -0400] "GET /api/v1/users/USERM/goals/LPM/datapoints.json?access_token=REDACTED&count=7&sort=daystamp HTTP/2.0" 200 2398 "-" "BeeSwift/6.0 (com.beeminder.beeminder; build:19; iOS 15.6.1) Alamofire/4.9.1" request_time="0.225" upstream_connect_time="0.008" upstream_header_time="0.224" upstream_response_time="0.224" upstream_cache_status="-"
200.107.24.229 - - [01/Oct/2022:14:14:58 -0400] "GET /api/v1/users/USERM/goals/LPM/datapoints.json?access_token=REDACTED&count=7&sort=daystamp HTTP/2.0" 200 2398 "-" "BeeSwift/6.0 (com.beeminder.beeminder; build:19; iOS 15.6.1) Alamofire/4.9.1" request_time="0.310" upstream_connect_time="0.000" upstream_header_time="0.300" upstream_response_time="0.308" upstream_cache_status="-"
200.107.24.229 - - [01/Oct/2022:14:14:58 -0400] "GET /api/v1/users/USERM/goals/LPM/datapoints.json?access_token=REDACTED&count=7&sort=daystamp HTTP/2.0" 200 2398 "-" "BeeSwift/6.0 (com.beeminder.beeminder; build:19; iOS 15.6.1) Alamofire/4.9.1" request_time="0.301" upstream_connect_time="0.008" upstream_header_time="0.300" upstream_response_time="0.300" upstream_cache_status="-"
200.107.24.229 - - [01/Oct/2022:14:14:58 -0400] "GET /api/v1/users/USERM/goals/LPM/datapoints.json?access_token=REDACTED&count=7&sort=daystamp HTTP/2.0" 200 2398 "-" "BeeSwift/6.0 (com.beeminder.beeminder; build:19; iOS 15.6.1) Alamofire/4.9.1" request_time="0.332" upstream_connect_time="0.024" upstream_header_time="0.324" upstream_response_time="0.332" upstream_cache_status="-"
```

Some of these clients are requesting the same goal 90 times a second, and they don't even stop when we give them 50x errors.

**Comment by theospears** [comment]
@adamwolf  Could you do some quick aggregation on app/iOS version numbers we are seeing this from? I'm curious if BeeSwift 6.1 or a new iOS release have made this worse.

**Comment by adamwolf** [comment]
A glance this morning made it look like it's more common than it used to be, but the underlying retries have been happening for years.  I can run actual stats tomorrow.

Hypothesizing, if it has to do with Heatlhkit, and you fixed a bunch of healthkit issues in 6.1 (thanks!), that makes sense.

**Comment by adamwolf** [comment]
While waiting for access to App Store Connect et al, I started to gather stats for this, and ran into something.

The backend is serving up 304, Not Modified.  (It's probable they have empty bodies.)

A strict reading of RequestManager makes me think that returning a 304, depending upon how Alamofire unwraps it, may fail the whole request.  validate() only accepts 2xx status codes, for instance.

**Comment by adamwolf** [comment]
It doesn't seem like it can be validate, if it's happening with 200s as well.

**Comment by theospears** [comment]
It looks like this is resolved in 6.3.
